### PR TITLE
Support NoData in AutoWarped VRT

### PR DESCRIPTION
**Pull request by jwittner**
Main objective was to support no data when using the Auto Warped VRT by taking any src no data values and propagating them into the destination. This required some enhancements to resolving the working data type.

Along the way I resolved a few issues, tried to clean up things a little, and reduce some copy pasta.

**I'll completely understand if you want this broken into smaller PRs. I also expect that I should be adding some unit tests somewhere?** Any guidance is appreciated!

### CODE REVIEW

**Comment by jwittner** [comment]
Looks like I'm failing on a bunch of documentation warnings. Can anyone point me toward docs on how to run these builds so I can document my new methods and then confirm?

**Comment by rouault** [comment]
There are not only documentation warnings, but compiler errors/warnings
```
gdalwarper.cpp:1310: error: ‘nullptr’ was not declared in this scope --> we haven't yet switched to requiring C++11 although that might change soon
vrtwarped.cpp: In function ‘void* GDALAutoCreateWarpedVRT(void*, const char*, const char*, GDALResampleAlg, double, const GDALWarpOptions*)’:
vrtwarped.cpp:130: error: ISO C++ forbids declaration of ‘rasterBand’ with no type
vrtwarped.cpp:130: error: invalid conversion from ‘void*’ to ‘int’
vrtwarped.cpp:133: error: invalid conversion from ‘int’ to ‘void*’
vrtwarped.cpp:133: error:   initializing argument 1 of ‘double GDALGetRasterNoDataValue(void*, int*)’
vrtwarped.cpp:144: error: ‘nullptr’ was not declared in this scope
```

As far as documentation warnings, this is a matter of installing Doxygen and running "make docs 2>&1 | grep -i warnings" to see if Doxygen emit warnings

**Comment by jwittner** [comment]
Yep, saw those too! Thanks for the info on the documentation, I'll be looking into all of these next week.

**Comment by jwittner** [comment]
@rouault Just to be sure the below is the test that failed that is why my travis build failed yes?
```
 ------------ Failures ------------
Script: gcore/vsicurl.py
  TEST: vsicurl_test_redirect ... fail
    line 321: fail
 ----------------------------------
```

That test succeeds for me on Windows. Looking at the line:
```
f = gdal.VSIFOpenL('/vsicurl/http://localhost:%d/test_redirect/test.bin' % gdaltest.webserver_port, 'rb')
```
Digging into the VSIFOpen api I'm not sure how I could have affected this.

File: gdal/gcore/gdal_misc.cpp [hunk 1]
```diff
@@ -77,132 +77,163 @@ GDALDataType CPL_STDCALL
...
+    const bool bSigned = pabSigned[0] || pabSigned[1];
+    const bool bFloating =
+        CPL_TO_BOOL( GDALDataTypeIsFloating( eType1 ) ) ||
+        CPL_TO_BOOL( GDALDataTypeIsFloating( eType2 ) );
+    const bool bComplex = 
+        CPL_TO_BOOL( GDALDataTypeIsComplex( eType1 ) ) ||
+        CPL_TO_BOOL( GDALDataTypeIsComplex( eType2 ) );
+
+    const int nBits = pabSigned[0] != pabSigned[1] ?
+         std::max(panBits[pabSigned[1]], 2 * panBits[!pabSigned[1]]) :
```
Comment (rouault) [inline gdal/gcore/gdal_misc.cpp]: I have trouble understanding this logic. And this doesn't like to be symmetrical if you switch eType1 and eType2. Or at least this is not immediately obvious to my eyes

**Comment by rouault** [comment]
> Just to be sure the below is the test that failed that is why my travis build failed yes?

This is a "known" random failure. Ignore it (or if you wish to investigate what causes it that would be warmly welcome !)

File: gdal/gcore/gdal_misc.cpp (hunk 1 above)
Comment (jwittner) [inline gdal/gcore/gdal_misc.cpp]: Agreed!  I'll refactor.

**Comment by jwittner** [comment]
Ok. 

Is there a Trac issue for the test failure?

**Comment by rouault** [comment]
> Is there a Trac issue for the test failure?

I don't think there's one yet

File: gdal/gcore/gdal_misc.cpp (hunk 1 above)
Comment (jwittner) [inline gdal/gcore/gdal_misc.cpp]: Clarified this piece of code.

File: gdal/gcore/gdal_misc.cpp [hunk 2]
```diff
@@ -77,132 +90,161 @@ GDALDataType CPL_STDCALL
...
+ * @return a best fit GDALDataType for supporting the value 
+ */
+GDALDataType CPL_STDCALL GDALFindDataTypeForValue( 
+    double dValue, int bComplex )
+{
+    const bool bFloating = round(dValue) != dValue;
+    const bool bSigned = bFloating || dValue < 0;
+    const int nBits = GetMinBitsForValue(dValue);
+
+    CPLDebug("Application", "GDALFindDataTypeForValue - bFloating %d", bFloating );
```
Comment (jwittner) [inline gdal/gcore/gdal_misc.cpp]: I should erase these debug prints.

**Comment by rouault** [comment]
Excellent contribution ! I've just committed it (with removal of the debug traces you pointed) New test case would be indeed appreciated.

**Comment by jwittner** [comment]
Yay! Thanks a ton @rouault!

**Comment by schwehr** [comment]
Were the behavior changes with GdalMiscTest.GdalDataTypeUnion  in gdal_misc intended?  

https://trac.osgeo.org/gdal/ticket/6879

**Comment by rouault** [comment]
@jwittner Mind reviewing the report in https://trac.osgeo.org/gdal/ticket/6879 ? There seems to be at least a few regressions due to your changes

**Comment by jwittner** [comment]
Just saw these comments - will definitely look into it on Monday.

**Comment by jwittner** [comment]
@rouault Looking to expose GDALDataTypeUnion to Python in order to port the gunit tests over. Looking at the python SWIG documentation it doesn't look like I can run those on Windows?

How do we feel about moving the gunit tests into the repo? I've never used gunit.

In the meantime I'll dig into fixing this up.

**Comment by jwittner** [comment]
@rouault - Oh, it looks like there are c++ unit tests already. Are those run with the python tests?

**Comment by rouault** [comment]
You can add tests in atotest/cpp/test_cpl.cpp and run gdal_unit_test. Those are generally run in the Travvis CI setups before running the python tests

**Comment by rouault** [comment]
hum actually this should rather be in autotest/cpp/test_gdal.cpp

**Comment by jwittner** [comment]
Ok, the errant unions should be fixed here #217.

**Pull request closed without merging**